Handle negative indexes

bigodel opened the pull request:
This PR allows the user to specify negative indexes when getting values from the response. That means that, given the following response JSON

```json
{
  "foo": {
    "test": "Hello, World!",
    "nested": {"x": 99}
  },
  "bar": [42, 100, true, false, null],
  "empty-array":  [],
  "empty-object": {}
}
```

The user could access the last element of the `"bar"` key with `(verb-json-get (oref verb-last body) "bar" -1)`. For the second to last, the user would call `(verb-json-get (oref verb-last body) "bar" -2)` and so forth.

If the requested negative index tries to access an element that isn't on the bounds, such as trying to access index -6 of `"bar"` in our example, an error is thrown just as it would for positive indexes.

bigodel commented:
i have also taken the liberty to clean up and improve some functions that i ended up stumbling upon, but i could always create a new PR with those changes and some more, in due time.

federicotdn commented on verb.el:
```diff
@@ -836,15 +840,14 @@ Return t if there was a heading to move towards to and nil otherwise."
-              ;; 2) Take the (key . value) for each of those properties here
-              (mapcar (lambda (key) (cons (upcase key)
-                                          (org-entry-get (point) key
-                                                         'selective)))
-                      ;; 1) Get all doc properties and filter them by prefix
-                      (seq-filter (lambda (s) (string-prefix-p prefix s t))
-                                  (org-buffer-property-keys)))))
+  (thread-last
+    (org-buffer-property-keys)
+    ;; 1) Get all doc properties and filter them by prefix
+    (seq-filter (lambda (s) (string-prefix-p prefix s t)))
+    ;; 2) Get the value for each of those properties and return an alist
+    (mapcar (lambda (key) (cons (upcase key) (org-entry-get (point) key 'selective))))
+    ;; 3) Discard all (key . nil) elements in the list
+    (seq-filter #'cdr)))
```
Could you provide context on these changes?

bigodel replied:
i'm simply using the threading macro, introduced in 25.1, to make it more readable. also, `(seq-filter #'cdr list)` has the outcome of removing nils, since `(equal '(foo) '(foo . nil))`

federicotdn commented on verb.el:
```diff
@@ -1086,17 +1089,19 @@ been set once with `verb-var', and then prompt for VALUE.  Otherwise,
-                                                        (symbol-name (car e)))
-                                                      verb--vars))))
-         (val (or value (read-string (format "Set value for %s: " v))))
-         (elem (assoc-string v verb--vars)))
-    (when (string-empty-p v)
+  (let* ((name (or (and (stringp var) var)
+                   (ignore-error 'wrong-type-argument (symbol-name var))
+                   (completing-read "Variable: " (mapcar (lambda (e)
+                                                           (symbol-name (car e)))
+                                                         verb--vars))))
+         (key (intern name))
+         (val (or value (read-string (format "Set value for %s: " name))))
+         (elem (assq key verb--vars)))
+    (when (string-empty-p name)
       (user-error "%s" "Variable name can't be empty"))
     (if elem
         (setcdr elem val)
-      (push (cons (intern v) val) verb--vars))))
+      (push (cons key val) verb--vars))))
```
These ones too.

bigodel replied:
this allows you to provide a symbol as a variable name instead of just strings. i thought it would make sense since it will be stored as a symbol in the end. i was caught by surprise when i realized that the variable name needed to be a string

federicotdn commented on test/verb-test.el:
```diff
@@ -2210,19 +2210,26 @@
   (should (= (verb-json-get test-json "bar" 0) 42))
   (should (= (verb-json-get test-json "bar" 1) 100))
 
+  (should-not (= (verb-json-get test-json "bar" -1) nil))
+  (should (eq (verb-json-get test-json "bar" -2) :json-false))
+  (should (= (verb-json-get test-json "bar" -5)
+             (verb-json-get test-json "bar" 0)))
+  (should-error (verb-json-get test-json "bar" -6))
+  (should-error (verb-json-get test-json "bar" -999))
+
```
Thanks for adding tests!

bigodel replied:
least i could do :D

federicotdn reviewed (changes requested):
Hey! Apologies for the delay. I liked your idea, I was just wondering about some additional changes I saw on the commit.

Would you mind also doing the indentation changes to `verb-test.el` in a separate PR? Thank you!

bigodel commented:
> Hey! Apologies for the delay. I liked your idea, I was just wondering about some additional changes I saw on the commit.

i should be the one apologizing :P. somehow i wasn't notified of your response, and only recently have i came back here to check on the status. i have addressed your comments

> Would you mind also doing the indentation changes to verb-test.el in a separate PR? Thank you!

sure! i reverted it here

federicotdn reviewed (approved):
Looks very good! I'll merge it

federicotdn commented:
@bigodel Looks like there's a linting error, should be easy to fix:
```
emacs --batch -l vendor/package-lint/package-lint.el \
		 -f package-lint-batch-and-exit "verb.el"
Entering directory '/home/runner/work/verb/verb/'
verb.el:1093:20: error: You should depend on (emacs "27.1") if you need `ignore-error'.
```

federicotdn commented on verb.el:
```diff
@@ -1086,17 +1089,19 @@ been set once with `verb-var', and then prompt for VALUE.  Otherwise,
-         (elem (assoc-string v verb--vars)))
-    (when (string-empty-p v)
+  (let* ((name (or (and (stringp var) var)
+                   (ignore-error 'wrong-type-argument (symbol-name var))
```
I'll change this line via the Github UI:
```suggestion
                   (and (symbolp var) (symbol-name var))
```

bigodel commented:
> Looks very good! I'll merge it

thank you!
